### PR TITLE
Erc20 pay for tx fee

### PR DESCRIPTION
**Pull request by zjb0807**
Closes: #857
Closes: #1892

### CODE REVIEW

File: runtime/integration-tests/src/evm.rs [hunk 1]
```diff
@@ -706,3 +731,316 @@ fn test_evm_accounts_module() {
...
+			));
+
+			assert_eq!(
+				Currencies::free_balance(NATIVE_CURRENCY, &sub_account),
+				5 * dollar(NATIVE_CURRENCY)
+			);
+			// erc20 minimum_balance is 0
+			assert_eq!(
+				Currencies::free_balance(CurrencyId::Erc20(erc20_address_0()), &sub_account),
+				0
```
Comment (zqhxuyuan) [inline runtime/integration-tests/src/evm.rs:891]: is it dex pair not register to asset-registry cause ED of dex = 0?

https://github.com/AcalaNetwork/Acala/blob/8560555fb9572517dafcc6557f4ae1879fc99e1d/runtime/karura/src/lib.rs?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L788-L791

**Review by xlc — DISMISSED** [review]

File: modules/evm-bridge/src/mock.rs [hunk 2]
```diff
@@ -188,8 +188,14 @@ pub fn bob_evm_addr() -> EvmAddress {
 	EvmAddress::from_str("1000000000000000000000000000000000000002").unwrap()
 }
 
+pub fn alice_balance() -> u128 {
+	100_000_000_000_000_000_000_000u128
+}
```
Comment (shaunxw) [inline modules/evm-bridge/src/mock.rs:193]: Same here.

File: modules/transaction-payment/src/lib.rs [hunk 3]
```diff
@@ -387,24 +396,40 @@ pub mod module {
...
+	///
+	/// PoolSize: map CurrencyId => Balance
 	#[pallet::storage]
 	#[pallet::getter(fn pool_size)]
 	pub type PoolSize<T: Config> = StorageMap<_, Twox64Concat, CurrencyId, Balance, ValueQuery>;
 
 	/// The exchange rate between the given currency and native token.
 	/// This value is updated when upon swap from dex.
+	///
+	/// TokenExchangeRate: map CurrencyId => Ratio
```
Comment (shaunxw) [inline modules/transaction-payment/src/lib.rs:425]:
```suggestion
	/// TokenExchangeRate: map CurrencyId => Option<Ratio>
```

File: modules/transaction-payment/src/lib.rs [hunk 4]
```diff
@@ -497,6 +522,54 @@ pub mod module {
...
+		#[transactional]
+		pub fn set_global_fee_swap_path(origin: OriginFor<T>, fee_swap_path: Vec<CurrencyId>) -> DispatchResult {
+			T::UpdateOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				fee_swap_path.len() > 1
+					&& fee_swap_path[0] != T::NativeCurrencyId::get()
+					&& fee_swap_path[fee_swap_path.len() - 1] == T::NativeCurrencyId::get(),
+				Error::<T>::InvalidSwapPath
+			);
```
Comment (shaunxw) [inline modules/transaction-payment/src/lib.rs:536]:
```suggestion
			ensure!(
				fee_swap_path.len() > 1
					&& fee_swap_path.get(0) != Some(T::NativeCurrencyId::get())
					&& fee_swap_path.get(fee_swap_path.len() - 1) == Some(T::NativeCurrencyId::get()),
				Error::<T>::InvalidSwapPath
			);
```

File: modules/transaction-payment/src/lib.rs [hunk 5]
```diff
@@ -497,6 +522,54 @@ pub mod module {
...
+
+			GlobalFeeSwapPath::<T>::try_mutate(fee_swap_path[0], |maybe_path| -> DispatchResult {
```
Comment (shaunxw) [inline modules/transaction-payment/src/lib.rs:538]:
```suggestion
			GlobalFeeSwapPath::<T>::try_mutate(fee_swap_path.get(0).expect("ensured path not empty; qed"), |maybe_path| -> DispatchResult {
```

File: modules/transaction-payment/src/lib.rs [hunk 6]
```diff
@@ -497,6 +522,54 @@ pub mod module {
...
+					new_fee_swap_path: fee_swap_path,
+				});
+				Ok(())
+			})
+		}
+
+		/// Unset global fee swap path
+		#[pallet::weight(<T as Config>::WeightInfo::unset_global_fee_swap_path())]
+		#[transactional]
+		pub fn unset_global_fee_swap_path(origin: OriginFor<T>, currency_id: CurrencyId) -> DispatchResult {
```
Comment (shaunxw) [inline modules/transaction-payment/src/lib.rs:558]:
```suggestion
		pub fn remove_global_fee_swap_path(origin: OriginFor<T>, currency_id: CurrencyId) -> DispatchResult {
```

File: modules/transaction-payment/src/lib.rs [hunk 7]
```diff
@@ -793,24 +867,38 @@ where
 
 	/// Get trading path by user.
 	fn get_trading_path(who: &T::AccountId) -> Vec<Vec<CurrencyId>> {
-		let mut default_fee_swap_path_list = T::DefaultFeeSwapPathList::get();
-		if let Some(trading_path) = AlternativeFeeSwapPath::<T>::get(who) {
-			default_fee_swap_path_list.insert(0, trading_path.into_inner())
+		let mut trading_path: Vec<Vec<CurrencyId>> = Vec::new();
+
+		if let Some(path) = AlternativeFeeSwapPath::<T>::get(who) {
+			trading_path.insert(0, path.into_inner());
```
Comment (shaunxw) [inline modules/transaction-payment/src/lib.rs:873]:
```suggestion
			trading_path.push(path.into_inner());
```

File: modules/transaction-payment/src/lib.rs [hunk 8]
```diff
@@ -793,24 +867,38 @@ where
...
 		}
-		default_fee_swap_path_list
+
+		let global_fee_swap_path = GlobalFeeSwapPath::<T>::iter_values()
+			.map(|v| v.into_inner())
+			.collect::<Vec<_>>();
+		if !global_fee_swap_path.len().is_zero() {
+			trading_path.extend_from_slice(&global_fee_swap_path);
```
Comment (shaunxw) [inline modules/transaction-payment/src/lib.rs:880]:
```suggestion
			trading_path.extend(global_fee_swap_path);
```

File: modules/transaction-payment/src/lib.rs [hunk 9]
```diff
@@ -793,24 +867,38 @@ where
...
+		}
+
+		trading_path.extend_from_slice(&T::DefaultFeeSwapPathList::get());
```
Comment (shaunxw) [inline modules/transaction-payment/src/lib.rs:883]:
```suggestion
		trading_path.extend(T::DefaultFeeSwapPathList::get());
```

File: modules/transaction-payment/src/lib.rs [hunk 10]
```diff
@@ -387,24 +396,40 @@ pub mod module {
...
 
 	/// The alternative fee swap path of accounts.
+	///
+	/// AlternativeFeeSwapPath: map AccountId => Option<Vec<CurrencyId>>
 	#[pallet::storage]
 	#[pallet::getter(fn alternative_fee_swap_path)]
 	pub type AlternativeFeeSwapPath<T: Config> =
 		StorageMap<_, Twox64Concat, T::AccountId, BoundedVec<CurrencyId, T::TradingPathLimit>, OptionQuery>;
 
+	/// The global fee swap path.
```
Comment (shaunxw) [inline modules/transaction-payment/src/lib.rs:406]: will be helpful to document path priorities: alternative > global > default

**Review by shaunxw — APPROVED** [review]
Need to redo benchmark.

**Review by shaunxw — DISMISSED** [review]

**Pull request merged**